// ts-runtime/mutators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Mutator_1 = require("./mutators/Mutator");
exports.Mutator = Mutator_1.Mutator;
var ArrowFunctionMutator_1 = require("./mutators/ArrowFunctionMutator");
exports.ArrowFunctionMutator = ArrowFunctionMutator_1.ArrowFunctionMutator;
var AsExpressionMutator_1 = require("./mutators/AsExpressionMutator");
exports.AsExpressionMutator = AsExpressionMutator_1.AsExpressionMutator;
var BinaryExpressionMutator_1 = require("./mutators/BinaryExpressionMutator");
exports.BinaryExpressionMutator = BinaryExpressionMutator_1.BinaryExpressionMutator;
var BlockLikeMutator_1 = require("./mutators/BlockLikeMutator");
exports.BlockLikeMutator = BlockLikeMutator_1.BlockLikeMutator;
var ClassDeclarationMutator_1 = require("./mutators/ClassDeclarationMutator");
exports.ClassDeclarationMutator = ClassDeclarationMutator_1.ClassDeclarationMutator;
var FunctionDeclarationMutator_1 = require("./mutators/FunctionDeclarationMutator");
exports.FunctionDeclarationMutator = FunctionDeclarationMutator_1.FunctionDeclarationMutator;
var FunctionExpressionMutator_1 = require("./mutators/FunctionExpressionMutator");
exports.FunctionExpressionMutator = FunctionExpressionMutator_1.FunctionExpressionMutator;
var InterfaceDeclararionMutator_1 = require("./mutators/InterfaceDeclararionMutator");
exports.InterfaceDeclarationMutator = InterfaceDeclararionMutator_1.InterfaceDeclarationMutator;
var SourceFileMutator_1 = require("./mutators/SourceFileMutator");
exports.SourceFileMutator = SourceFileMutator_1.SourceFileMutator;
var TypeAliasDeclararionMutator_1 = require("./mutators/TypeAliasDeclararionMutator");
exports.TypeAliasDeclarationMutator = TypeAliasDeclararionMutator_1.TypeAliasDeclarationMutator;
var VariableDeclarationListMutator_1 = require("./mutators/VariableDeclarationListMutator");
exports.VariableDeclarationListMutator = VariableDeclarationListMutator_1.VariableDeclarationListMutator;
exports.mutators = [
    ArrowFunctionMutator_1.ArrowFunctionMutator,
    AsExpressionMutator_1.AsExpressionMutator,
    BinaryExpressionMutator_1.BinaryExpressionMutator,
    BlockLikeMutator_1.BlockLikeMutator,
    ClassDeclarationMutator_1.ClassDeclarationMutator,
    FunctionDeclarationMutator_1.FunctionDeclarationMutator,
    FunctionExpressionMutator_1.FunctionExpressionMutator,
    InterfaceDeclararionMutator_1.InterfaceDeclarationMutator,
    SourceFileMutator_1.SourceFileMutator,
    TypeAliasDeclararionMutator_1.TypeAliasDeclarationMutator,
    VariableDeclarationListMutator_1.VariableDeclarationListMutator,
];
function getMutators() {
    var instances = [];
    for (var _i = 0, mutators_1 = exports.mutators; _i < mutators_1.length; _i++) {
        var mutator = mutators_1[_i];
        instances.push(new mutator());
    }
    return instances;
}
exports.getMutators = getMutators;
